python/dgl/contrib/sampling/dis_sampler.py

```diff
@@ -0,0 +1,88 @@
+# This file contains distributed samplers.
+from ...node_flow import NodeFlow
+from ...network import _create_sender, _create_receiver
+from ...network import _send_subgraph, _recv_subgraph
+from ...network import _finalize_sender, _finalize_receiver
+
+class Sender(object):
+    """The Sender class for distributed sampler.
+
+    Users use the Sender class to send sampled subgraph 
+    to remote trainer machine.
+
+    Parameters
+    ----------
+    ip : ip address of remote machine
+    port : listen port of remote machine
+    """
+    def __init__(self, ip, port):
+        self._ip = ip
+        self._port = port
+        self._sender = _create_sender(ip, port)
+
+    def Send(self, nodeflow):
+        """Send nodeflow to remote receiver
+
+        Parameter
+        ---------
+        nodeflow : a NodeFlow object
+        """
+        _send_subgraph(self._sender, nodeflow)
+
+    def BatchSend(self, nodeflow_list):
+        """Send a batch of nodeflow to remote receiver
+
+        Parameter
+        ---------
+        nodeflow_list : a list of nodeflow object
+        """
+        _batch_send_subgraph(self._sender, nodeflow_list)
+
+    def Finalize(self):
+        """Finalize Sender
+        """
+        _finalize_sender(self._sender)
```

Comment:
When the `Finalize` is called? In other word, why not put in `__del__`? Unless you have following scenario:
```python
sender = Sender(...)
# do some sending stuff
sender.Finalize()
sender.Restart(...)  # however, I didn't see any restart functions here.
```

Reply:
Finalize() will send a special msg to Receiver node, and Receive node can clean up the corresponding backend threads when a Sender finishes its job. If user don't invoke this method, nothing will happen in sender. But it is a waste of thread for Receive node.

Reply:
I think it's more flexible to add a Finalize() API to users to control the Sender.

Reply:
I don't know if we have the scenario, which we should invoke the Finalize() API and then start a new Sender, and these two Sender have the same life cycle.

Reply:
If `Finalize` is the right function to be called, then my point is to invoke it automatically once the sender object is out-of-scope. This makes it easier to handle exceptions and resources can be correctly released. For example:
```python
with SamplerSender(...) as sender:
  # do a  lot of stuff
# sender goes out-of-scope, __del__ function will call Finalize.
```

Reply:
Ok, I know your concern. If this is the only scenario we may meet in DGL, that's no problem to put `finalize` into `__del__`.